src/server/conn.rs

```diff
@@ -74,31 +76,33 @@ pub(super) struct SpawnAll<I, S> {
 ///
 /// Polling this future will drive HTTP forward.
 #[must_use = "futures do nothing unless polled"]
-pub struct Connection<I, S>
+pub struct Connection<T, S>
 where
     S: Service,
+    T: AsyncRead + AsyncWrite,
 {
-    pub(super) conn: Either<
+    pub(super) conn: Option<
+        Either<
         proto::h1::Dispatcher<
             proto::h1::dispatch::Server<S>,
             S::ResBody,
-            I,
+            Rewind<T>,
```

Comment:
This is just a thought, feel free to tell me it's crazy: it may be possible to actually just pass `T` here, and not need `Rewind<T>` *only for h1*. It seems like the only time `rewind` is called is when deconstructing h1 and building an h2 dispatcher, so in `try_h2` is where it could be `let io = Rewind::new(io); io.rewind(read_buf); ...`. I have a hunch this will mean that this new feature doesn't affect the performance of HTTP1 at all, since the `read` calls would never need to check the rewind buffer.

Reply:
Yeah, this is something I thought about, I cant remember why I didn't do this, let me try to look through the code and jog my memory.